File: python/easy/70_climbing_stairs.py
"""
70. Climbing Stairs
Easy

You are climbing a staircase. It takes n steps to reach the top.

Each time you can either climb 1 or 2 steps. In how many distinct ways can
you climb to the top?
"""

d = {1: 1, 2: 2}


class Solution:
    def climbStairs(self, n: int) -> int:
        if n in d:
            return d[n]

        if n - 1 not in d:
            d[n - 1] = self.climbStairs(n - 1)
        if n - 2 not in d:
            d[n - 2] = self.climbStairs(n - 2)

        return d[n - 1] + d[n - 2]

File: go/70_climbing_stairs.go
// 70. Climbing Stairs
// Easy
//
// You are climbing a staircase. It takes n steps to reach the top.
//
// Each time you can either climb 1 or 2 steps. In how many distinct ways can
// you climb to the top?
package main

var m map[int]int = map[int]int{1: 1, 2: 2}

func climbStairs(n int) int {
	if n, ok := m[n]; ok {
		return n
	}

	m[n-1] = climbStairs(n - 1)
	m[n-2] = climbStairs(n - 2)

	return m[n-1] + m[n-2]
}
